collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py

```diff
@@ -1,58 +1,76 @@
+
+
+def eval_iter_callback(tensors, global_vars):
+    if "all_preds" not in global_vars.keys():
+        global_vars["all_preds"] = []
+    if "all_labels" not in global_vars.keys():
+        global_vars["all_labels"] = []
+    if "all_subtokens_mask" not in global_vars.keys():
+        global_vars["all_subtokens_mask"] = []
+
+    all_subtokens_mask, all_logits, all_labels = [], [], []
 
     for kv, v in tensors.items():
-        if 'logits' in kv:
+        if kv.startswith('logits'):
             for v_tensor in v:
                 for logit_tensor in v_tensor:
-                    logits_lists.append(logit_tensor.detach().cpu().tolist())
+                    all_logits.append(tensor2list(logit_tensor))
+
+        if kv.startswith('labels'):
+            for v_tensor in v:
+                for label_tensor in v_tensor:
+                    all_labels.extend(tensor2list(label_tensor))
 
-        if 'seq_ids' in kv:
+        if kv.startswith('subtokens_mask'):
             for v_tensor in v:
-                for seq_id_tensor in v_tensor:
-                    seq_ids.append(seq_id_tensor.detach().cpu().tolist())
+                for subtokens_mask_tensor in v_tensor:
+                    all_subtokens_mask.extend(
+                        tensor2list(subtokens_mask_tensor))
 
-    correct_labels, incorrect_labels, correct_preds, total_preds, \
-        total_correct = eval_data_layer.eval_preds(logits_lists, seq_ids)
+    all_preds = list(np.argmax(np.asarray(all_logits), 2).flatten())
+    global_vars["all_preds"].extend(all_preds)
+    global_vars["all_labels"].extend(all_labels)
+    global_vars["all_subtokens_mask"].extend(all_subtokens_mask)
 
-    global_vars["correct_labels"] += correct_labels
-    global_vars["incorrect_labels"] += incorrect_labels
-    global_vars["correct_preds"] += correct_preds
-    global_vars["total_preds"] += total_preds
-    global_vars["total_correct"] += total_correct
 
+def list2str(l):
+    return ' '.join([str(j) for j in l])
 
-def eval_epochs_done_callback(global_vars):
 
-    correct_labels = global_vars["correct_labels"]
-    incorrect_labels = global_vars["incorrect_labels"]
-    correct_preds = global_vars["correct_preds"]
-    total_preds = global_vars["total_preds"]
-    total_correct = global_vars["total_correct"]
+def eval_epochs_done_callback(global_vars, label_ids, none_label_id=0):
+    labels = np.asarray(global_vars['all_labels'])
+    preds = np.asarray(global_vars['all_preds'])
+    subtokens_mask = np.asarray(global_vars['all_subtokens_mask'])
 
-    accuracy = correct_labels / (correct_labels + incorrect_labels)
+    labels = labels[subtokens_mask]
+    preds = preds[subtokens_mask]
 
-    p = correct_preds / total_preds
-    r = correct_preds / total_correct
-    f1 = 2 * p * r / (p + r)
+    accuracy = sum(labels == preds) / labels.shape[0]
+    logger.info(f'Accuracy: {accuracy}')
 
-    print(f"Accuracy = {accuracy}")
-    print(f"F1= {f1}")
+    i = 0
+    if preds.shape[0] > 21:
```

Comment:
where does 21 come from?

Reply:
added comments and introduced a variable for the magic number